editor/components/inserter/menu.js

```diff
@@ -29,32 +77,206 @@ class InserterMenu extends wp.element.Component {
 		return () => {
 			this.props.onInsertBlock( slug );
 			this.props.onSelect();
-			this.setState( { filterValue: '' } );
+			this.setState( {
+				filterValue: '',
+				currentFocus: null
+			} );
 		};
 	}
 
-	render() {
-		const { position = 'top' } = this.props;
-		const blocks = wp.blocks.getBlocks();
-		const isShownBlock = block => block.title.toLowerCase().indexOf( this.state.filterValue.toLowerCase() ) !== -1;
-		const blocksByCategory = blocks.reduce( ( groups, block ) => {
-			if ( ! isShownBlock( block ) ) {
-				return groups;
-			}
-			if ( ! groups[ block.category ] ) {
-				groups[ block.category ] = [];
+	getVisibleBlocks( blockTypes ) {
+		return blockTypes.filter( this.isShownBlock );
+	}
+
+	sortBlocksByCategory( blockTypes ) {
+		const getCategoryIndex = ( item ) => {
+			return this.categories.findIndex( ( category ) => category.slug === item.slug );
+		};
+
+		return blockTypes.sort( ( a, b ) => {
+			return getCategoryIndex( a ) - getCategoryIndex( b );
+		} );
+	}
+
+	groupByCategory( blockTypes ) {
+		return blockTypes.reduce( ( accumulator, block ) => {
+			// If already an array push block on else add array of block.
+			if ( Array.isArray( accumulator[ block.category ] ) ) {
+				accumulator[ block.category ].push( block );
+				return accumulator;
 			}
-			groups[ block.category ].push( block );
-			return groups;
+
+			accumulator[ block.category ] = [ block ];
+			return accumulator;
 		}, {} );
-		const categories = wp.blocks.getCategories();
+	}
+
+	getVisibleBlocksByCategory( blockTypes ) {
+		return flow(
+			this.getVisibleBlocks,
+			this.sortBlocksByCategory,
+			this.groupByCategory
+		)( blockTypes );
+	}
+
+	findNext( currentBlock, blockTypes ) {
+		/**
+		 * null is the value that will trigger iterating back to
+		 * the top of the list of block types.
+		 */
+		if ( null === currentBlock ) {
+			return blockTypes[ 0 ].slug;
+		}
+
+		const currentIndex = blockTypes.findIndex( ( blockType ) => currentBlock === blockType.slug );
```

Comment:
We don't use the Babel polyfill currently, meaning we don't have access to ES2015 base type instance methods like `Array#findIndex`, `String#startsWith`, etc. Static members like `Object.assign` and `Array.from` are fair game however.

We could decide to use the polyfill, but it adds non-trivial weight to the build. The developer experience implications could make it worth including.